### PR TITLE
feat(metrics): Add support for 'tuple' conditions [TET-319]

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR adds the support for the `tuple` function  in the `where` filtering on `MetricsQuery`. The support for this function should allow the indexing of tag keys and their respective tag values.

An example usage of a `Condition` involving `tuple`s:
```python
...
where=[
                Condition(
                    lhs=Function(
                        function="tuple",
                        parameters=[
                            Column(
                                name="tags[transaction]",
                            )
                        ],
                    ),
                    op=Op.IN,
                    rhs=Function(
                        function="tuple",
                        parameters=[("/foo",), ("/bar",)],
                    ),
                )
            ],
...
```

### CODE REVIEW

File: tests/sentry/snuba/metrics/test_query_builder.py [hunk 1]
```diff
@@ -1036,3 +1036,114 @@ def test_valid_latest_release_alias_filter(self):
...
+
+    def test_resolve_tags_with_binary_tuple(self):
+        tags = [("/foo", "ios"), ("/bar", "android")]
+
+        indexer.record(self.use_case_id, self.org_id, string="transaction")
+        indexer.record(self.use_case_id, self.org_id, string="platform")
+
+        for transaction, platform in tags:
+            indexer.record(use_case_id=self.use_case_id, org_id=self.org_id, string=transaction)
+            indexer.record(use_case_id=self.use_case_id, org_id=self.org_id, string=platform)
```
Comment (ahmedetefy) [inline tests/sentry/snuba/metrics/test_query_builder.py:1101]: Can you combine those maybe like this? 
```suggestion
        for transaction, platform in [("transaction", "platform")] + tags:
            indexer.record(use_case_id=self.use_case_id, org_id=self.org_id, string=transaction)
            indexer.record(use_case_id=self.use_case_id, org_id=self.org_id, string=platform)
```

**Review by ahmedetefy — COMMENTED** [review]
Overall looks good, just minor comments

File: src/sentry/snuba/metrics/query_builder.py [hunk 2]
```diff
@@ -103,8 +103,14 @@ def resolve_tags(
...
-        # drop unknown strings:
-        return [x for x in elements if x != STRING_NOT_FOUND]
+        # drop unknown strings.
+        filtered_elements = [x for x in elements if x != STRING_NOT_FOUND]
+        # We check whether it is a list or tuple in order to know which type to return. This is needed
+        # because in the "tuple" function the parameters must be a list of tuples and not a list of lists.
+        if isinstance(input_, list):
+            return filtered_elements
+        elif isinstance(input_, tuple):
+            return tuple(filtered_elements)
```
[the remark this answers is not in the document]
Comment (iambriccardo) [reply src/sentry/snuba/metrics/query_builder.py:113]: Thanks! Always looking for tips on syntax sugar

[the remark this answers is not in the document]
Comment (ahmedetefy) [reply src/sentry/snuba/metrics/query_builder.py:113]: This is not about syntactic sugar tho. this is about behaviour change

File: tests/sentry/snuba/metrics/test_query_builder.py (hunk 1 above)
Comment (ahmedetefy) [reply tests/sentry/snuba/metrics/test_query_builder.py:1101]: How does it achieve better readability?

**Review by ahmedetefy — APPROVED** [review]

**Pull request merged**